### ISSUE TITLE
Linker Script File Proposal

### ISSUE THREAD

**Issue by ReinhardKeil** [Open-CMSIS-Pack/devtools#655]
Please provide feedback for https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/projmgr/docs/Manual/Linker%20Script%20File%20(Proposal).md in this issue

**Comment by ReinhardKeil** [comment]
WiP linker scripts are here:
https://github.com/ARM-software/CMSIS-View/tree/eventstatistic/Examples/EventStatistic/RTE/Device/ARMCM3

Commands to generate the linker scripts are here:
https://github.com/ARM-software/CMSIS-View/blob/1cc05a27f917c31df0b617eb5c4ac11f6a1621cb/Examples/EventStatistic/build.py#L177-L184

**Comment by ReinhardKeil** [comment]
@JonatanAntoni could you extend https://github.com/ARM-software/CMSIS-View/blob/eventstatistic/Examples/EventStatistic/RTE/Device/ARMCM3/memory_layout.h to also provide a NOINIT attribute for RAM.

**Comment by JonatanAntoni** [comment]
NOINIT is handled "implicitly" by assigning data to `.bss.noinit` section. This removes the manual effort for e.g. EventRecorder to explicitly create a NOINIT memory region and assign EventRecorder buffer into it.

**Comment by ReinhardKeil** [comment]
OK, so this is handled at the level of the C source module instead?  Does that work across all toolchains?  Do we have provisions in CMSIS-Core for that already?

**Comment by JonatanAntoni** [comment]
This is currently work in progress:
https://github.com/ARM-software/CMSIS-View/blob/0ba0aab9208374a4f938de2091c7fd34ea4e955f/EventRecorder/Source/EventRecorder.c#L42-L55

Once this is confirmed to work we can move it to CMSIS-Core (cmsis_compiler.h).

**Comment by JonatanAntoni** [comment]
@mario-pi, It's a feature request which is not yet implemented but planned.

**Comment by mario-pi** [comment]
@JonatanAntoni thanks for the clarification. Is the #667 issue related to the current implementation instead?

**Comment by JonatanAntoni** [comment]
@mario-pi, yes, without #667 implemented, csolution won't properly pass any .icf file to IAR Linker command line.

**Comment by JonatanAntoni** [comment]
@tarek-bochkati, may I ask you to have a look into the example I prepared at [https://github.com/ARM-software/CMSIS-View/blob/eventstatistic/Examples/EventStatistic](https://github.com/ARM-software/CMSIS-View/blob/eventstatistic/Examples/EventStatistic/RTE/Device/ARMCM3/memory_layout.h), please?

The IAR configuration compiles without errors but doesn't produce expected output (stdio and event recorder log file) via semihosting. I guess something is wrong and the example runs into some sort of fault.

**Comment by jkrech** [comment]
Due to the different linker script file formats a toolchain independent project requires one file per `compiler`.
```
    - group: LinkerScripts
      files:
        - file: ./AC6.linkerscript.sct
          for-compiler: AC6
        - file: ./IAR.linkerscript.scf
          for-compiler: IAR
        - file: ./GCC.linkerscript.ld
          for-compiler: GCC
```
The advantage of running a C-Preprocessor over the different linker scripts which share the same C-Preprocessor annoations ist the fact that they are all configured using a single header file.

A `<component>` description today specifies
```
<file category="linkerScript" name="Startup/iar_arm.icf" version="1.0.0" attr="config" condition="IAR"/>
<file category="linkerScript" name="Startup/ac6_arm.sct" version="1.0.0" attr="config" condition="AC6"/>
<file category="linkerScript" name="Startup/gcc_arm.ld" version="2.1.0" attr="config" condition="GCC"/>
```
In order to support preprocessing we could make use of the to be implemented `language` attribute. I am proposing `c` as we need to run a C-preprocessor.
```
<file category="header" language="link" name="startup/regions.h" version="1.0.0" attr="config" />
<file category="linkerScript" language="c"  name="startup/iar_arm.icf" version="1.0.0" attr="config" condition="IAR"/>
<file category="linkerScript" language="c" name="startup/ac6_arm.sct" version="1.0.0" attr="config" condition="AC6"/>
<file category="linkerScript" language="c" name="startup/gcc_arm.ld" version="2.1.0" attr="config" condition="GCC"/>
```

Further feedback to consider:
- People do not like the fact that files related to the linker are copied as configuration files to their project, which are not used.
- linkerScript files can come from multiple selected components. How to find them in order to explicitly add them to the project for selection?
- should we require dedicated components for Linker (Cclass="Compiler" Cgroup="Linker")?

Do we define a dedicated directory "Linker" to hold linker files for the user to choose from?
Should we allow a component reference in the `linker:` node?

Fallback to generic linker scripts shipped as part of cmsis-toolbox `./etc` directory with `regions.h` file generated by `csolution` (similar to `RTE_Components`) based on `<device>` and `<board>` `<memory>` information.

Note: The 'regions.h' header file is provided as 'pre-include' file. This allows to preprocess with different include file names without the need to change the `#include` statement.

**Comment by jkrech** [comment]
A first version of the implementation will be available in CMSIS-Toolbox 1.7.0 and we will iterate over it.
If there is any ideas for managing linker scripts for Dual Core projects and Secure / Non-secure Trustzone projects, proposals  are welcome. The current implementation only covers simple projects/devices.
